WordPress/openverse
# Simplify and consolidate photon log handling, improve tallying

## Fixes



Fixes #2225 by @sarayourfriend

## Description



This PR simplifies and consolidates the photon exception handling by making a few changes:

- If a status code is returned, it is always tallied under `thumbnail_respone_code:{month}:{response_code}` (e.g. `thumbnail_response_code:2023-05:200`, for global response code counts) and `thumbnail_response_code_by_domain:{domain}:{month}:{response_code}` (e.g. `thumbnail_response_code_by_domain:sample.subdomain.org:2023-05:200`, for per-domain counts)
- If an exception is raised, it is always tallied under `thumbnail_error:{exception_name}:{domain}:{month}` (e.g. `thumbnail_error:requests.SSLError:sample.subdomain.org:2023-05`)
- If an HTTP error specifically is raised via the `raise_for_status`, it is _additionally_ tallied under `thumbnail_http_error:{domain}:{month}:{status_code}:{error_text}` (e.g. `thumbnail_http_error:sample.subdomain.org:2023-05:400:Cannot process request`)
- Every exception raised will return a `UpstreamThumbnailException` of the same form
- All initial errors for a particular generic exception (per domain) will be sent to sentry up to the initial threshold (default 100), then every repeated threshold (default 1000) after that.

This has a few upshots:
- The exception handling code is more straightforward
- We are tallying multiple pieces of information per-request, regardless of status
- In practice, an issue that continues to be alerted will alert at higher frequency at the top of the month when a new key is generated (and thus the threshold resets)

I believe this addresses all of the concerns raised in #2225, save for `If status codes overlap various messages from Photon, disambiguate them with an additional token (i.e., 400:invalid_params and 400:ssl etc).` - I could not easily find different error messages which photon might emit besides `We cannot complete this request, remote data could not be fetched`, so I'm hoping that the addition of `error_text` in the redis key will help with differentiating this. I didn't want to presume/prescribe a mapping in case that missed other cases, so that seemed like the easiest catch-all.


## Testing Instructions


I don't know that there's a way to test this locally with the error cases we're seeing in production, since that would require hooking the local API up to the staging/production database. Tests have changed drastically so they should pass. You can also add an intentional exception within the Photon "get" function and check that redis has the values, for example:

1. Add a `raise ValueError("Foo")` after the `requests.get` call and other increments but before the `raise_for_status` call
2. Launch the API and run a search, then open a thumbnail result
3. Run `redis-cli -p 50263` to connect to redis
4. Observe keys like the following: 

```
127.0.0.1:50263> keys thumbnail_*
1) "thumbnail_response_code:2023-05:200"
2) "thumbnail_response_code_by_domain:cdn.stocksnap.io:2023-05:200"
3) "thumbnail_error:builtins.ValueError:cdn.stocksnap.io:2023-05"
```
5. Query for the key results and see that they are counting occurrences:

```
127.0.0.1:50263> get "thumbnail_response_code_by_domain:cdn.stocksnap.io:2023-05:200"
"1"
```


## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
